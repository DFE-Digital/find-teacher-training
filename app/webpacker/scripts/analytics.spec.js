import { initFormAnalytics, initExternalLinkAnalytics, initNavigationAnalytics } from "./analytics"

global.ga = jest.fn()

describe("Analytics", () => {
  afterEach(() => {
    global.ga.mockClear()
  })

  describe("initFormAnalytics", () => {
    beforeEach(() => {
      // Note: not a <form> because JSDOM throws a fit https://github.com/jsdom/jsdom/issues/1937
      document.body.innerHTML = `
        <div data-ga-event-form="Some form">
          <input type="checkbox" data-ga-event-form-input="A ticked item" checked>
          <input type="checkbox" data-ga-event-form-input="An unticked item">
          <input type="checkbox" data-ga-event-form-input="Another ticked item" checked>
          <input type="submit">
        </div>
        `
      initFormAnalytics()
    })

    it("triggers correct GA event when users submit with checked checkboxes", () => {
      document.querySelector("[type=submit]").click()
      expect(ga.mock.calls).toMatchSnapshot()
    })
  })

  describe("initExternalLinkAnalytics", () => {
    beforeEach(() => {
      document.body.innerHTML = `
        <div>
          <a href="http://example.com"></a>
          <a href="https://google.com"></a>

          <a href="http://localhost/foo"></a>
          <a href="/foo"></a>
          <a href="#foo"></a>
        </div>
        `
      initExternalLinkAnalytics()
    })

    it("triggers correct GA events when users click on external links", () => {
      Array.from(document.querySelectorAll("a")).forEach($el => $el.click())
      expect(ga.mock.calls).toMatchSnapshot()
    })
  })

  describe("initNavigationAnalytics", () => {
    beforeEach(() => {
      document.body.innerHTML = `
        <div data-ga-event-navigation="Some navigation">
          <a href="#about"></a>
          <a href="http://example.com"></a>
        </div>
        `
      initNavigationAnalytics()
    })

    it("triggers correct GA events when users click on navigation links", () => {
      Array.from(document.querySelectorAll("a")).forEach($el => $el.click())
      expect(ga.mock.calls).toMatchSnapshot()
    })
  })
})
