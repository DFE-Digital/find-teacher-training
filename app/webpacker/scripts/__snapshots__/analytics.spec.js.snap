// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics initExternalLinkAnalytics triggers correct GA events when users click on external links 1`] = `
Array [
  Array [
    "send",
    "event",
    Object {
      "eventAction": "http://example.com",
      "eventCategory": "External Link Clicked",
      "transport": "beacon",
    },
  ],
  Array [
    "send",
    "event",
    Object {
      "eventAction": "https://google.com",
      "eventCategory": "External Link Clicked",
      "transport": "beacon",
    },
  ],
]
`;

exports[`Analytics initFormAnalytics triggers correct GA event when users submit with checked checkboxes 1`] = `
Array [
  Array [
    "send",
    "event",
    Object {
      "eventAction": "A ticked item, Another ticked item",
      "eventCategory": "Form: Some form",
      "transport": "beacon",
    },
  ],
]
`;

exports[`Analytics initNavigationAnalytics triggers correct GA events when users click on navigation links 1`] = `
Array [
  Array [
    "send",
    "event",
    Object {
      "eventAction": "#about",
      "eventCategory": "Some navigation Navigation Link Clicked",
      "transport": "beacon",
    },
  ],
  Array [
    "send",
    "event",
    Object {
      "eventAction": "http://example.com",
      "eventCategory": "Some navigation Navigation Link Clicked",
      "transport": "beacon",
    },
  ],
]
`;
